NXP LPSPI: Fixing word sizes, SPI_HOLD_ON_CS flag and optimizing interrupts in CPU-based transfers

decsny opened the pull request:
Fix word sizes other than 8 bits on LPSPI driver and optimize for less interrupts.  Fixes bugs that I find while working on https://github.com/zephyrproject-rtos/zephyr/pull/87838 .  Also combines the logging modules

Commit:
spi_nxp_lpspi: Optimize TX fill for less interrupt

Optimize the TX fill algorithm to have less interrupts by filling the TX
fifo as much as possible during each interrupt handle.

Before, the algorithm was just a very simple, fill the TX fifo with as
much from only the current buffer as possible, then send it and wait for
the next interrupt. Now the algorithm is to fill the TX fifo as much as
possible, even if it means reading from multiple buffers during the
interrupt.

This has the advantage from master mode of having less interrupts. And
it is very important for slave mode because the slave mode does not
control the pacing of the transfer and so therefore should fill as much
as possible whenever possible in order not to miss a deadline.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Support word size < 8

The LPSPI does support word sizes such as 6 or 7, anything as small as 2
bits. So fix the checks and the math to allow for this in the driver.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Fix word size > 8

Fix calculations for larger than 8 bit word sizes

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Use one logging module

Use one logging module for LPSPI driver instead of 3

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

hakehuang commented:
> @hakehuang please help support regression testing this PR

@decsny nice optimize. and now only below platforms need you attentions

- frdm_rw612/rw612
- rd_rw612_bga
- mimxrt1180_evk/mimxrt1189/cm33

the get the same log as below

```
*** Booting Zephyr OS build v4.0.0-rc2 ***
Hello World! frdm_rw612/rw612
*** Booting Zephyr OS build v4.1.0-2445-g246f02b09460 ***
SPI test on buffers TX/RX 0x10044b80/0x10044b60, frame size = 8
Polling...Running TESTSUITE spi_extra_api_features
===================================================================
START - test_spi_lock_release
 PASS - test_spi_lock_release in 0.001 seconds
===================================================================
TESTSUITE spi_extra_api_features succeeded
Running TESTSUITE spi_extra_api_features
===================================================================
START - test_spi_lock_release
 PASS - test_spi_lock_release in 0.001 seconds
===================================================================
TESTSUITE spi_extra_api_features succeeded
Running TESTSUITE spi_loopback
===================================================================
Testing loopback spec: SLOW
START - test_nop_nil_bufs
 PASS - test_nop_nil_bufs in 0.001 seconds
===================================================================
START - test_spi_async_call
 PASS - test_spi_async_call in 0.133 seconds
===================================================================
START - test_spi_complete_large_transfers
 PASS - test_spi_complete_large_transfers in 0.132 seconds
===================================================================
START - test_spi_complete_loop
 PASS - test_spi_complete_loop in 0.001 seconds
===================================================================
START - test_spi_complete_multiple
 PASS - test_spi_complete_multiple in 0.001 seconds
===================================================================
START - test_spi_null_rx_buf_set
ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/lib/libc/picolibc/assert.c:27
E: r0/a1:  0x00000004  r1/a2:  0x0000001b  r2/a3:  0x00000006
E: r3/a4:  0x00000004 r12/ip:  0x00000008 r14/lr:  0x1800381d
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x180090b2
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x10044c38 (test_spi_null_rx_buf_set)
E: Halting system
```

decsny commented:
> > @hakehuang please help support regression testing this PR
> 
> @decsny nice optimization. and now only below platforms need you attentions
> 
>     * frdm_rw612/rw612
> 
>     * rd_rw612_bga
> 
>     * mimxrt1180_evk/mimxrt1189/cm33
> 
> 
> the get the rd series log as below
> 
> ```
> *** Booting Zephyr OS build v4.0.0-rc2 ***
> Hello World! frdm_rw612/rw612
> *** Booting Zephyr OS build v4.1.0-2445-g246f02b09460 ***
> SPI test on buffers TX/RX 0x10044b80/0x10044b60, frame size = 8
> Polling...Running TESTSUITE spi_extra_api_features
> ===================================================================
> START - test_spi_lock_release
>  PASS - test_spi_lock_release in 0.001 seconds
> ===================================================================
> TESTSUITE spi_extra_api_features succeeded
> Running TESTSUITE spi_extra_api_features
> ===================================================================
> START - test_spi_lock_release
>  PASS - test_spi_lock_release in 0.001 seconds
> ===================================================================
> TESTSUITE spi_extra_api_features succeeded
> Running TESTSUITE spi_loopback
> ===================================================================
> Testing loopback spec: SLOW
> START - test_nop_nil_bufs
>  PASS - test_nop_nil_bufs in 0.001 seconds
> ===================================================================
> START - test_spi_async_call
>  PASS - test_spi_async_call in 0.133 seconds
> ===================================================================
> START - test_spi_complete_large_transfers
>  PASS - test_spi_complete_large_transfers in 0.132 seconds
> ===================================================================
> START - test_spi_complete_loop
>  PASS - test_spi_complete_loop in 0.001 seconds
> ===================================================================
> START - test_spi_complete_multiple
>  PASS - test_spi_complete_multiple in 0.001 seconds
> ===================================================================
> START - test_spi_null_rx_buf_set
> ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/lib/libc/picolibc/assert.c:27
> E: r0/a1:  0x00000004  r1/a2:  0x0000001b  r2/a3:  0x00000006
> E: r3/a4:  0x00000004 r12/ip:  0x00000008 r14/lr:  0x1800381d
> E:  xpsr:  0x21000000
> E: Faulting instruction address (r15/pc): 0x180090b2
> E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> E: Current thread: 0x10044c38 (test_spi_null_rx_buf_set)
> E: Halting system
> ```
> 
> and rt1180 series get below failure
> 
> ```
> 2025-04-16 13:57:05,610 - twister - DEBUG - DEVICE: �*** Booting Zephyr OS build v4.1.0-2445-g246f02b09460 ***
> 2025-04-16 13:57:05,619 - twister - DEBUG - DEVICE: SPI test on buffers TX/RX 0x30004320/0x30004300, frame size = 8, DMA enabled (without CONFIG_NOCACHE_MEMORY)
> 2025-04-16 13:57:05,624 - twister - DEBUG - DEVICE: Polling...Running TESTSUITE spi_extra_api_features
> 2025-04-16 13:57:05,630 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:05,632 - twister - DEBUG - DEVICE: START - test_spi_lock_release
> 2025-04-16 13:57:05,838 - twister - DEBUG - DEVICE: E: Timeout waiting for transfer complete
> 2025-04-16 13:57:05,850 - twister - DEBUG - DEVICE: Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:109: spi_loopback_transceive: (ret is true)
> 2025-04-16 13:57:05,853 - twister - DEBUG - DEVICE: SPI transceive failed, code -116
> 2025-04-16 13:57:05,857 - twister - DEBUG - DEVICE: FAIL - test_spi_lock_release in 0.221 seconds
> 2025-04-16 13:57:05,863 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:05,866 - twister - DEBUG - DEVICE: TESTSUITE spi_extra_api_features failed.
> 2025-04-16 13:57:05,870 - twister - DEBUG - DEVICE: Running TESTSUITE spi_extra_api_features
> 2025-04-16 13:57:05,876 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:05,879 - twister - DEBUG - DEVICE: START - test_spi_lock_release
> 2025-04-16 13:57:06,083 - twister - DEBUG - DEVICE: E: Timeout waiting for transfer complete
> 2025-04-16 13:57:06,095 - twister - DEBUG - DEVICE: Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:109: spi_loopback_transceive: (ret is true)
> 2025-04-16 13:57:06,098 - twister - DEBUG - DEVICE: SPI transceive failed, code -116
> 2025-04-16 13:57:06,102 - twister - DEBUG - DEVICE: FAIL - test_spi_lock_release in 0.220 seconds
> 2025-04-16 13:57:06,108 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:06,112 - twister - DEBUG - DEVICE: TESTSUITE spi_extra_api_features failed.
> 2025-04-16 13:57:06,114 - twister - DEBUG - DEVICE: Running TESTSUITE spi_loopback
> 2025-04-16 13:57:06,120 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:06,123 - twister - DEBUG - DEVICE: Testing loopback spec: SLOW
> 2025-04-16 13:57:06,125 - twister - DEBUG - DEVICE: START - test_nop_nil_bufs
> 2025-04-16 13:57:06,129 - twister - DEBUG - DEVICE: PASS - test_nop_nil_bufs in 0.001 seconds
> 2025-04-16 13:57:06,135 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:06,138 - twister - DEBUG - DEVICE: START - test_spi_async_call
> 2025-04-16 13:57:06,810 - twister - DEBUG - -- west flash: using runner linkserver
> ```

it looks like the failure is from a new test case was added on main a couple days ago, I can look into fixing it on this PR

decsny commented:
> > @hakehuang please help support regression testing this PR
> 
> @decsny nice optimization. and now only below platforms need you attentions
> 
>     * frdm_rw612/rw612
> 
>     * rd_rw612_bga
> 
>     * mimxrt1180_evk/mimxrt1189/cm33
> 
> 
> the get the rd series log as below
> 
> ```
> *** Booting Zephyr OS build v4.0.0-rc2 ***
> Hello World! frdm_rw612/rw612
> *** Booting Zephyr OS build v4.1.0-2445-g246f02b09460 ***
> SPI test on buffers TX/RX 0x10044b80/0x10044b60, frame size = 8
> Polling...Running TESTSUITE spi_extra_api_features
> ===================================================================
> START - test_spi_lock_release
>  PASS - test_spi_lock_release in 0.001 seconds
> ===================================================================
> TESTSUITE spi_extra_api_features succeeded
> Running TESTSUITE spi_extra_api_features
> ===================================================================
> START - test_spi_lock_release
>  PASS - test_spi_lock_release in 0.001 seconds
> ===================================================================
> TESTSUITE spi_extra_api_features succeeded
> Running TESTSUITE spi_loopback
> ===================================================================
> Testing loopback spec: SLOW
> START - test_nop_nil_bufs
>  PASS - test_nop_nil_bufs in 0.001 seconds
> ===================================================================
> START - test_spi_async_call
>  PASS - test_spi_async_call in 0.133 seconds
> ===================================================================
> START - test_spi_complete_large_transfers
>  PASS - test_spi_complete_large_transfers in 0.132 seconds
> ===================================================================
> START - test_spi_complete_loop
>  PASS - test_spi_complete_loop in 0.001 seconds
> ===================================================================
> START - test_spi_complete_multiple
>  PASS - test_spi_complete_multiple in 0.001 seconds
> ===================================================================
> START - test_spi_null_rx_buf_set
> ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/lib/libc/picolibc/assert.c:27
> E: r0/a1:  0x00000004  r1/a2:  0x0000001b  r2/a3:  0x00000006
> E: r3/a4:  0x00000004 r12/ip:  0x00000008 r14/lr:  0x1800381d
> E:  xpsr:  0x21000000
> E: Faulting instruction address (r15/pc): 0x180090b2
> E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> E: Current thread: 0x10044c38 (test_spi_null_rx_buf_set)
> E: Halting system
> ```
> 
> and rt1180 series get below failure
> 
> ```
> 2025-04-16 13:57:05,610 - twister - DEBUG - DEVICE: �*** Booting Zephyr OS build v4.1.0-2445-g246f02b09460 ***
> 2025-04-16 13:57:05,619 - twister - DEBUG - DEVICE: SPI test on buffers TX/RX 0x30004320/0x30004300, frame size = 8, DMA enabled (without CONFIG_NOCACHE_MEMORY)
> 2025-04-16 13:57:05,624 - twister - DEBUG - DEVICE: Polling...Running TESTSUITE spi_extra_api_features
> 2025-04-16 13:57:05,630 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:05,632 - twister - DEBUG - DEVICE: START - test_spi_lock_release
> 2025-04-16 13:57:05,838 - twister - DEBUG - DEVICE: E: Timeout waiting for transfer complete
> 2025-04-16 13:57:05,850 - twister - DEBUG - DEVICE: Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:109: spi_loopback_transceive: (ret is true)
> 2025-04-16 13:57:05,853 - twister - DEBUG - DEVICE: SPI transceive failed, code -116
> 2025-04-16 13:57:05,857 - twister - DEBUG - DEVICE: FAIL - test_spi_lock_release in 0.221 seconds
> 2025-04-16 13:57:05,863 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:05,866 - twister - DEBUG - DEVICE: TESTSUITE spi_extra_api_features failed.
> 2025-04-16 13:57:05,870 - twister - DEBUG - DEVICE: Running TESTSUITE spi_extra_api_features
> 2025-04-16 13:57:05,876 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:05,879 - twister - DEBUG - DEVICE: START - test_spi_lock_release
> 2025-04-16 13:57:06,083 - twister - DEBUG - DEVICE: E: Timeout waiting for transfer complete
> 2025-04-16 13:57:06,095 - twister - DEBUG - DEVICE: Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:109: spi_loopback_transceive: (ret is true)
> 2025-04-16 13:57:06,098 - twister - DEBUG - DEVICE: SPI transceive failed, code -116
> 2025-04-16 13:57:06,102 - twister - DEBUG - DEVICE: FAIL - test_spi_lock_release in 0.220 seconds
> 2025-04-16 13:57:06,108 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:06,112 - twister - DEBUG - DEVICE: TESTSUITE spi_extra_api_features failed.
> 2025-04-16 13:57:06,114 - twister - DEBUG - DEVICE: Running TESTSUITE spi_loopback
> 2025-04-16 13:57:06,120 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:06,123 - twister - DEBUG - DEVICE: Testing loopback spec: SLOW
> 2025-04-16 13:57:06,125 - twister - DEBUG - DEVICE: START - test_nop_nil_bufs
> 2025-04-16 13:57:06,129 - twister - DEBUG - DEVICE: PASS - test_nop_nil_bufs in 0.001 seconds
> 2025-04-16 13:57:06,135 - twister - DEBUG - DEVICE: ===================================================================
> 2025-04-16 13:57:06,138 - twister - DEBUG - DEVICE: START - test_spi_async_call
> 2025-04-16 13:57:06,810 - twister - DEBUG - -- west flash: using runner linkserver
> ```

hi @hakehuang , the RW612 does not have an LPSPI, so irrelevant to this PR, I can look into RT1180 though, I never tried that platform before

hakehuang commented:
> hi @hakehuang , the RW612 does not have an LPSPI, so irrelevant to this PR, I can look into RT1180 though, I never tried that platform before

Thanks checking @decsny , I just realize RD RW series issue has the same root cause as https://github.com/zephyrproject-rtos/zephyr/issues/88704

Commit:
spi_nxp_lpspi: Support SPI_HOLD_ON_CS FLAG

Support SPI_HOLD_ON_CS flag in the CPU-based driver. To do this we will
set CONTC bit to continue previous command. Technically it may not be
necessary right now, and could just not clear CONT bit...
but in the future in the lpspi driver we
will decouple the config/init of a transfer from the SDK
and therefore have more control over TCR,
and when we write the TCR, we need to take CONTC bit into account
otherwise a new command will be made. So this approach is how
it should be handled in the driver going forward in my opinion, even
if it might be possible without this bit right now, I want to introduce
it's usage now.

This commit also does a minor refactor in the ISR and adds some comments
to make the strange CS behavior and strange handling code more clear to
future readers.

Also, make the early predicted SPI xfer end code only happen for spi
versions where it is necessary, since I think that code is really the
best we can do but might have a race condition, where possible the last
word is not finished sending when we end the xfer. So limit the
potential affect to v1 lpspi where the workaround is actually required
due to stalling behavior.

Lastly, set the LPSPI into master mode at active low in init, due to
it being the most common case, we want the SPI CS lines to be
initialized at init of driver. I don't think it's worth it to make it
configurable at this time, but in the future it could be if needed.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

hakehuang reviewed (approved):
board testing PASS
